plugins/ingest/src/main/java/org/elasticsearch/ingest/Pipeline.java

```diff
@@ -65,7 +65,7 @@ public Builder(String name) {
             this.name = name;
         }
 
-        public Builder(Map<String, Object> config) {
+        public Builder(Map<String, Object> config, Map<String, Processor.Builder> processorRegistry) {
```

Comment:
well.. as I said, **IMO** the separation is cleaner. But if you keep the prototype way, you'll eventually need to keep the current `build()` method and introduce a new `Processor build(Map)` method **next** to it... it's a bit messy tbh... the same applies to the `Pipeline.Builder`.

Another option is leave things as they are (so leave the current `void fromMap(Map)` method) and then introduce a factory to a builder... however you look at it, we need a factory somewhere... a construct that is registered by "type" that can create the appropriate builder/processor for that type... something along the lines of:

```
Processor.Builder builder = registry.get("type").create();
```

Another option is to mandate all builders to have an empty ctor and use reflection to create those. So instead of injecting a set of builders, you'd work with a set of builder classes and then the code will just create new instances using reflection on default ctor.